emma/OntoEmma.py

```diff
@@ -533,17 +584,208 @@ def _align_string_equiv(s_kb, t_kb):
+        indices = hmod.compute()
+        print("Length of assignment: %i" % len(indices))
+
+        for row, column in indices:
+            row_ent_id = s_kb.entities[row].research_entity_id
+            col_ent_id = t_kb.entities[column].research_entity_id
+            score = sim_scores[(row_ent_id, col_ent_id)]
+            if score >= constants.SIM_SCORE_THRESHOLD:
+                alignment.append((s_kb.entities[row].research_entity_id,
+                                  t_kb.entities[column].research_entity_id,
+                                  score))
+        return alignment
+
+    def _compute_alignment(self, align_strat, l_scores, n_scores, s_kb, t_kb):
+        """
+        Compute best bipartite alignment between s_kb and t_kb based on scores
+        :param l_scores: entity similarity scores
+        :param n_scores: entity similarity scores updated based on similarity of neighborhood entities
+        :param s_kb:
+        :param t_kb:
+        :return:
+        """
+        if align_strat == "best":
+            return self._apply_best_alignment_strategy(n_scores)
+        elif align_strat == "all":
+            return self._apply_all_alignment_strategy(n_scores)
+        elif align_strat == "modh":
+            return self._apply_modh_alignment_strategy(n_scores, s_kb, t_kb)
+        else:
+            raise NotImplementedError("Error: Unknown alignment strategy.")
+
+    def _compute_neighborhood_similarities(self, scores, s_kb, t_kb, iterations=0):
+        """
+        Compute neighborhood similarities of each pair of entities based on entity similarity scores
+        :param scores: Entity similarity scores
+        :param s_kb: Source KB
+        :param t_kb: Target KB
+        :param iterations: number of iterations for propagating neighborhood similarity
+        :return:
+        """
+        updated_neighborhood_sim = copy(scores)
+
+        # iteratively calculate global similarity scores
+        for i in range(0, iterations):
+            neighborhood_sim = dict()
+
+            # iterate through all alignments
+            for (s_ent_id, t_ent_id), score in tqdm.tqdm(updated_neighborhood_sim.items(),
+                                                         total=len(updated_neighborhood_sim)):
+                s_ent = s_kb.get_entity_by_research_entity_id(s_ent_id)
+                t_ent = t_kb.get_entity_by_research_entity_id(t_ent_id)
+
+                # generate regions around s_ent and t_ent not included s_ent and t_ent
+                s_region = self._get_region_around_ent(s_ent, s_kb)
+                t_region = self._get_region_around_ent(t_ent, t_kb)
+
+                # sum regional contributions to similarity
+                neighborhood_sum = 0.0
+                distance_weights = 0.0
+                for s_neighbor_id, t_neighbor_id in itertools.product(s_region, t_region):
+                    if len(s_region[s_neighbor_id]) == len(t_region[t_neighbor_id]) and \
```

Comment:
I think this condition makes sure the two neighbors have the same distance, but that's not clear in the code. Consider adding a comment explaining what an example of `s_region[s_neighbor_id]` might look like.